app/Main.hs

```diff
@@ -23,61 +33,82 @@ data Command = MoveRight
 tokenize :: String -> [Command]
-tokenize [] = []
-tokenize (x:xs)
-  | x == '>' = MoveRight : tokenize xs
-  | x == '<' = MoveLeft : tokenize xs
-  | x == '+' = IncCell: tokenize xs
-  | x == '-' = DecCell : tokenize xs
-  | x == '.' = PrintCell : tokenize xs
-  | x == ',' = InputCell: tokenize xs
-  | x == '[' = OpenBracket : tokenize xs
-  | x == ']' = CloseBracket: tokenize xs
-  | otherwise = tokenize xs
+tokenize = catMaybes . fmap charToCommand
+  where
+    charToCommand c =
+      case c of
+        '>' -> Just MoveRight
+        '<' -> Just MoveLeft
+        '+' -> Just IncCell
+        '-' -> Just DecCell
+        '.' -> Just PrintCell
+        ',' -> Just InputCell
+        '[' -> Just OpenBracket
+        ']' -> Just CloseBracket
+        _ -> Nothing
 
-
--- Sets a specific cell to a value
 setCell :: (Int -> Int) -> CellState -> Int -> CellState
-setCell f cs i = take i cs ++ [if j < 0 then 255 else j] ++ drop (i + 1) cs
-    where j = f $ cs !! i
+setCell _ [] _ = []
+setCell f (c:rest) 0 = f c : rest
+setCell f (c:rest) n = c : setCell f rest (n-1)
 
 incCell :: CellState -> Int -> CellState
 incCell = setCell (+1)
 
 decCell :: CellState -> Int -> CellState
-decCell = setCell (\x -> x - 1)
-
+decCell = setCell (subtract 1)
 
 -- Parse
 -- Close bracket is different from the others as it
 -- doesn't bind itself to the parse process
 parse :: [Command] -> StateT MoveState IO [Command]
-parse [] = StateT (\x -> pure ([], x))
-parse (x:xs) = if x == CloseBracket then s else s >>= parse
-    where s = StateT (\v@(i, pc, cs) ->
-                if i < 0 then error $ show v
-                else
-                case x of
-                    MoveRight    -> pure (xs, (i + 1, pc ++ [x], cs))
-                    MoveLeft     -> pure (xs, (i - 1, pc ++ [x], cs))
-                    IncCell      -> pure (xs, (i, pc ++ [x], incCell cs i))
-                    DecCell      -> pure (xs, (i, pc ++ [x], decCell cs i))
-                    OpenBracket  -> do (xs', (i', pc', cs')) <- liftIO $ runStateT (parse xs) (i, [], cs)
-                                       pure (xs', (i, pc ++ [x] ++ pc', cs'))
-                    CloseBracket -> if cs !! i > 0
-                                    then
-                                       do (xs', ms') <- liftIO $ runStateT (parse (pc ++ [x] ++ xs)) (i, [], cs)
-                                          pure (xs', ms')
-                                    else pure (xs, (i, pc ++ [x], cs))
-                    InputCell    -> do c <- liftIO getChar
-                                       pure (xs, (i, pc ++ [x], setCell (const $ fromEnum c) cs i))
-                    PrintCell    -> do liftIO $ print (chr (cs !! i))
-                                       return (xs, (i, pc ++ [x], cs))
-            )
+parse [] = pure [] 
+parse (x:xs) =
+  case x of
+    CloseBracket -> s
+    _ -> s >>= parse
+  where
+    s = do
+      ms@(MoveState pos cmds cls) <- get 
+      when (pos < 0) . error $ show ms
```

Comment:
`when` does a monadic action only if the argument is true. Bit neater than `if .. then .. else`